src/zcash/NoteEncryption.cpp

```diff
@@ -101,12 +101,19 @@ void KDF(unsigned char K[NOTEENCRYPTION_CIPHER_KEYSIZE],
 
 namespace libzcash {
 
-boost::optional<SaplingNoteEncryption> SaplingNoteEncryption::FromDiversifier(diversifier_t d) {
+boost::optional<SaplingNoteEncryption> SaplingNoteEncryption::FromDiversifier(
+    diversifier_t d,
+    boost::optional<uint256> esk_provided
```

Comment:
I believe we only call `FromDiversifier` in two places: from `SaplingNotePlaintext::encrypt`, and from a test in `test_noteencryption.cpp`. The former always passes `esk`; the latter can be easily changed (and should be changed so that we're testing the case that is actually used).

```suggestion
    uint256 esk
```